src/compiler/cell_manager.rs

```diff
@@ -163,8 +197,27 @@ pub struct MaxWidthCellManager {
 
 impl CellManager for MaxWidthCellManager {
     fn place<F, StepArgs>(&self, unit: &mut CompilationUnit<F, StepArgs>) {
+        // If there exists shared signals, determine whether MaxWidthCellManager will return steps
+        // with different heights. If yes, panic.
+        // Can also use `same_height` method after placing all signals.
+        // However, checking at the front is more efficient by skipping placement if steps don't
+        // have the same height.
+        if !unit.shared_signals.is_empty() {
+            let max_width = self.max_width;
+            let mut signal_len_by_step_type = unit.step_types.values().map(|step_type| {
+                (step_type.signals.len() + unit.forward_signals.len() + unit.shared_signals.len())
+                    .div(max_width)
+                    + 1
+            });
+            let first = signal_len_by_step_type.next().unwrap();
+            if !signal_len_by_step_type.all(|len| len == first) {
```

Comment:
`signal_len_by_step_type` should be `step_height_by_step_type`. Sorry for the confusion.